### PR TITLE
Splitting istio charts into two parts deploying of CRDs and actual de…

### PR DESCRIPTION
**Pull request by sbezverk**
Since helm renders charts in non-deterministic order, in some cases CRD objects might not yet exist at the time when new objects based on CRD definitions get deployed, resulting in failed deployment. This PR splits CRD definitions from istio deployment charts. CRD deployment charts must run before the main istio deployment.

### CODE REVIEW

**Comment by sdake** [comment]
I gave this a spin and it consistently returns:
```
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes/helm$ helm install istio-common-objects/ --name istio-crds
Error: release istio-crds failed: namespaces "default" is forbidden: User "system:serviceaccount:kube-system:default" cannot get namespaces in the namespace "default"
```

Recommend running in a fresh cluster prior to testing.  An end to end test followup patch would be good too :)

**Comment by sdake** [comment]
@kyessenov sadly no.  Helm takes all files in the /templates directory, renders them, and creates them in parallel without regard to ordering.  At one time long ago we attempted adding ordering into helm charts using a tool called entrypoint which provides "choreography" to containers at runtime.  IMO this was and is the wrong solution.  Perhaps @technosophos has a solution in mind.  I think one thing that may work is providing specific ordering for specific resource types on instantation.  E.g. namespace first, CRDs second, etc.    There was another bug report (I'll have to hunt it down) on separating the CRDs from the rest of the manifests as they were causing race conditions I believe on IBM's cloud.

**Comment by sbezverk** [comment]
@kyessenov Unfortunately no, helm does not provide any way to synchronize deployment, what is worse is that the order of rendered charts is random and changes randomly.  So if there is a dependency on order, it has to be managed either by the explicit order of deployed charts or by the application by using some dependency mechanisms and restarts.

**Comment by sbezverk** [comment]
@sdake As I mentioned in the chat, you need to configure RBAC rule which normally would be done in a normally configured and operational cluster. Here is the example of a RBAC rule I use in my development cluster.
kubectl replace -f <(cat <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: cluster-admin
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: Group
  name: system:masters
- kind: Group
  name: system:authenticated
EOF
)

**Comment by sdake** [comment]
possibly related issue: https://github.com/istio/istio/pull/2017#issuecomment-351838612

File: install/kubernetes/helm/deployIstioHelmCharts.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+echo "Deploying Istio..."
+
+echo "Creating RBAC rules for Istio..."
+
+helm install istio-rbac --name istio-rbac
+retCode=$?
+if [ $retCode != '0' ]; then
+  echo "Creating of RBAC rules failed, exiting..."
+  exit 1
```
Comment (sdake) [inline install/kubernetes/helm/deployIstioHelmCharts.sh]: Knowing the exit code from helm is more important than knowing the step in which the Helm deployment failed.  That is indicated by the echo operation, however, the helm return code is never indicated.  Please change exit 1 to exit $retCode.

File: install/kubernetes/helm/deployIstioHelmCharts.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  exit 1 
+fi
+
+echo "Creating a namespace for Istio..."
+
+helm install istio-namespace --name istio-namespace
+retCode=$?
+if [ $retCode != '0' ]; then
+  echo "Creating of namespace failed, exiting..."
+  exit 2
```
Comment (sdake) [inline install/kubernetes/helm/deployIstioHelmCharts.sh]: see last comment.

File: install/kubernetes/helm/deployIstioHelmCharts.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+fi
+
+echo "Creating CRD objects for Istio..."
+
+helm install istio-crds --name istio-crds
+retCode=$?
+if [ $retCode != '0' ]; then
+  echo "Creating of CRDs objects failed, exiting..."
+  exit 3
```
Comment (sdake) [inline install/kubernetes/helm/deployIstioHelmCharts.sh]: see last comment.

File: install/kubernetes/helm/deployIstioHelmCharts.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+fi
+
+echo "Deploying Istio..."
+
+helm install istio --name istio
+retCode=$?
+if [ $retCode != '0' ]; then
+  echo "Deploying Istio failed, exiting..."
+  exit 4
```
Comment (sdake) [inline install/kubernetes/helm/deployIstioHelmCharts.sh]: see last comment.

File: install/kubernetes/helm/deployIstioHelmCharts.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+helm install istio-crds --name istio-crds
+retCode=$?
+if [ $retCode != '0' ]; then
+  echo "Creating of CRDs objects failed, exiting..."
+  exit 3
+fi
+
+echo "Deploying Istio..."
+
+helm install istio --name istio
```
Comment (sdake) [inline install/kubernetes/helm/deployIstioHelmCharts.sh]: I found myself consistently wanting to do the following:
helm install istio --name istio --set istio.ingress.use_nodeport=true

perhaps taking an argument to this last step in the deploy process would be helpful.

**Comment by sdake** [comment]
Patch appears to work (with bookinfo sample and modified helm install to use nodeport as described in the review: helm install istio --name istio --set istio.ingress.use_nodeport=true
).  Need charts for add-ons next ;-)  Few small comments in the deploy shell script to simply it's function.

**Comment by sdake** [comment]
Note this patch was tested with https://github.com/istio/istio/pull/2303 as a parent.

File: install/kubernetes/helm/deployIstioHelmCharts.sh (hunk 5 above)
Comment (sbezverk) [reply install/kubernetes/helm/deployIstioHelmCharts.sh]: I added helmExtraArgs parameter which can be used to pass to helm any additional arguments which must be recognizable by helm.
so in your case then you would use
deployIstioHelmCharts.sh "--set istio.ingress.use_nodeport=true"

File: install/kubernetes/helm/deployIstioHelmCharts.sh (hunk 4 above)
Comment (sbezverk) [reply install/kubernetes/helm/deployIstioHelmCharts.sh]: done

File: install/kubernetes/helm/deployIstioHelmCharts.sh (hunk 3 above)
Comment (sbezverk) [reply install/kubernetes/helm/deployIstioHelmCharts.sh]: done

File: install/kubernetes/helm/deployIstioHelmCharts.sh (hunk 2 above)
Comment (sbezverk) [reply install/kubernetes/helm/deployIstioHelmCharts.sh]: done

File: install/kubernetes/helm/deployIstioHelmCharts.sh (hunk 1 above)
Comment (sbezverk) [reply install/kubernetes/helm/deployIstioHelmCharts.sh]: done

**Comment by luisyonaldo** [comment]
I can't use this helm chart because I can't not pull from image repository:
```console
Failed to pull image "gcr.io/istio-testing/proxy:d82d983e983dc14fe4ce8b39374a91ef0ced6779": rpc error: code = Unknown desc = unauthorized: authentication required
```

**Comment by luisyonaldo** [comment]
also, for me the idea of using helm is not need to use any shell script to install manifest, no idea why using a install script instead of using helm hooks to install CRD and then the manifest that use them.

I already put everything on one helm chart and seems to be working, need to fix previous issue to be able to fully test it. If you want I can create a PR for this

**Comment by sbezverk** [comment]
@luisyonaldo I suspect your image tag is wrong, as a result the image pull fails. Could you try to do it manually to confirm? 
```console
docker pull gcr.io/istio-testing/proxy:d82d983e983dc14fe4ce8b39374a91ef0ced6779
```
As per having a single chart, as far as I know helm does not have any ordering hooks, if it was added, could you please point me to the doc, thanks a lot.

**Comment by luisyonaldo** [comment]
same issue
```
# docker pull gcr.io/istio-testing/proxy:d82d983e983dc14fe4ce8b39374a91ef0ced6779
Pulling repository gcr.io/istio-testing/proxy
unauthorized: authentication required
```
I also run the chart without any changes and I got same issue.. it works only for `initsializer` image but fails for `ingress`, `mixer` and `pilot`

**Comment by luisyonaldo** [comment]
https://docs.helm.sh/developing_charts/#hooks

I did the same as in `prometheus-operator` repo, you define CRD first then fire a post-install hoook that wait till the CRD are ready and later install the manifests that depend on the CRD.

you can check it in my branch https://github.com/istio/istio/compare/master...luisyonaldo:helm-refactor is currently in WIP

**Comment by sbezverk** [comment]
@luisyonaldo You need to run updateVersion.sh which will update values.yaml with the latest images' tags. Well at least it used to update, recently there were some changes done to updateVersion.sh I have not tested new version of it yet. Thanks a lot for the doc, I will go through it right away.

**Comment by sbezverk** [comment]
@luisyonaldo Could you please check the URL? I cannot access it.

**Comment by luisyonaldo** [comment]
the helm docs sites have been intermittent all daylong, try to go to [helm.sh](https://helm.sh/) and then go the docs. Right now seems to be down, only cached version works, but keep refreshing :)

**Comment by schnie** [comment]
Docs also here: https://github.com/kubernetes/helm/blob/master/docs/charts_hooks.md

**Comment by luisyonaldo** [comment]
@sbezverk can you check https://github.com/istio/istio/pull/2438?

**Pull request merged**